src/gauge/constructs/fat7LinkConstructs.h

```diff
@@ -41,6 +42,144 @@ template<class floatT,size_t HaloDepth,CompressionType comp>
+  else assert(0);
+
+  int nu_add = 0;
+  for (int nu_h = 0; nu_h < 3; nu_h++) {
+
+    if(nu_h == excluded_dir) nu_add = 1;
+    int nu = nu_h + nu_add; // we want nu here to go in ascending order; if excluded_dir == 1, then nu should go 0, 2, 3
+    if(nu == mu) continue;
+    //if(nu == excluded_dir) assert(0); //shouldn't occur
+
+    gSite downNu = GInd::site_dn(origin,nu);
+    gSite upNu = GInd::site_up(origin,nu);
+    gSite upMudownNu = GInd::site_dn(upMu, nu);
+    assert(gAcc_idx != nu_h);
+
+    if(nu_h == 0)temp_gAcc_nu_excluded_dir=gAcc_0;
+    else if(nu_h == 1)temp_gAcc_nu_excluded_dir=gAcc_1;
+    else if(nu_h == 2)temp_gAcc_nu_excluded_dir=gAcc_2;
+      
+    // nu > 0
+    temp += temp_gAcc_mu_excluded_dir.getLink(GInd::getSiteMu(origin,nu))*temp_gAcc_nu_excluded_dir.getLink(GInd::getSiteMu(upNu,mu))*temp_gAcc_mu_excluded_dir.getLinkDagger(GInd::getSiteMu(upMu,nu));
+    // nu < 0
+    temp += temp_gAcc_mu_excluded_dir.getLinkDagger(GInd::getSiteMu(downNu,nu))*temp_gAcc_nu_excluded_dir.getLink(GInd::getSiteMu(downNu,mu))*temp_gAcc_mu_excluded_dir.getLink(GInd::getSiteMu(upMudownNu,nu));
+  }
+  return temp;
+}
+
+template<class floatT,size_t HaloDepth,CompressionType comp>
+__host__ __device__ GSU3<floatT> inline hypThreeLinkStaple_third_level(gaugeAccessor<floatT,comp> gAcc_0, gaugeAccessor<floatT,comp> gAcc_1, gaugeAccessor<floatT,comp> gAcc_2, gaugeAccessor<floatT,comp> gAcc_3, gSiteMu siteMu, gaugeAccessor<floatT,comp> temp_gAcc_mu, gaugeAccessor<floatT,comp> temp_gAcc_nu) {
+
+  typedef GIndexer<All,HaloDepth> GInd;
+  GSU3<floatT> temp = gsu3_zero<floatT>();
+  GSU3<floatT> temp_chk = gsu3_zero<floatT>();
+  int mu = siteMu.mu;
+  gSite origin = GInd::getSite(siteMu.isite);
+  gSite upMu = GInd::site_up(origin,mu);
+
+  if(mu == 0)temp_gAcc_mu=gAcc_0;
+  else if(mu == 1)temp_gAcc_mu=gAcc_1;
+  else if(mu == 2)temp_gAcc_mu=gAcc_2;
+  else if(mu == 3)temp_gAcc_mu=gAcc_3;
+ 
+  for (int nu_h = 1; nu_h < 4; nu_h++) {
+    int nu = (mu+nu_h)%4;
+    gSite downNu = GInd::site_dn(origin,nu);
+    gSite upNu = GInd::site_up(origin,nu);
+    gSite upMudownNu = GInd::site_dn(upMu, nu);
+
+    if(nu == 0)temp_gAcc_nu=gAcc_0;
+    else if(nu == 1)temp_gAcc_nu=gAcc_1;
+    else if(nu == 2)temp_gAcc_nu=gAcc_2;
+    else if(nu == 3)temp_gAcc_nu=gAcc_3;
+    
+    // nu > 0
+    temp += temp_gAcc_mu.getLink(GInd::getSiteMu(origin,nu))*temp_gAcc_nu.getLink(GInd::getSiteMu(upNu,mu))*temp_gAcc_mu.getLinkDagger(GInd::getSiteMu(upMu,nu));
+    // nu < 0
+    temp += temp_gAcc_mu.getLinkDagger(GInd::getSiteMu(downNu,nu))*temp_gAcc_nu.getLink(GInd::getSiteMu(downNu,mu))*temp_gAcc_mu.getLink(GInd::getSiteMu(upMudownNu,nu));
+  }
+  assert(!(temp==temp_chk));
+  return temp;
+}
+
+//project to su3
+template<class floatT,size_t HaloDepth,CompressionType comp>
+__host__ __device__ GSU3<floatT> inline su3unitarize(gaugeAccessor<floatT,comp> gAcc, gaugeAccessor<floatT,comp> gAcc_base, gSiteMu siteMu) {
```

Comment:
Do we really need this function? If so, please explain somewhere what the difference is to the already existing unitarization functions available in  `Gaugefield<floatT,onDevice,HaloDepth,comp>::su3latunitarize()` in gaugefield_device.cpp? And also rename it properly.
Also if it is required, it shouldn't be placed in the fat7LinkConstructs.h file.

Reply:
The function was introduced to be consistent with the original implementation in the milc code. As we discussed in a previous meeting, the function was not to be removed as it is a different type of implementation, which might be used in the future. I personally have not made this myself, but as far as I remember, it is the difference between using the cross product in the first 2 rows to reconstruct the third, or to use all 3 rows and flow towards the unitary direction.

Reply:
Makes sense. However, this difference should be made clear in the code, since there are currently two unitarize functions with the same name but different algorithms. Otherwise this may cause confusion/problems in the future. Also it shouldn't be placed in the fat7LinkConstructs.h.